File: sdk/go/immortal.go
package immortal

import (
	"github.com/immortal-engine/immortal/internal/bus"
	"github.com/immortal-engine/immortal/internal/dna"
	"github.com/immortal-engine/immortal/internal/event"
	"github.com/immortal-engine/immortal/internal/healing"
)

type Config struct {
	Name      string
	Mode      string
	GhostMode bool
}

type App struct {
	config  Config
	bus     *bus.Bus
	healer  *healing.Healer
	dna     *dna.DNA
	running bool
}

func New(config Config) *App {
	if config.Name == "" {
		config.Name = "immortal-app"
	}
	if config.Mode == "" {
		config.Mode = "reactive"
	}
	h := healing.NewHealer()
	h.SetGhostMode(config.GhostMode)
	return &App{
		config: config,
		bus:    bus.New(),
		healer: h,
		dna:    dna.New(config.Name),
	}
}

func (a *App) Heal(rule healing.Rule) {
	a.healer.AddRule(rule)
}

func (a *App) Ingest(e *event.Event) {
	a.bus.Publish(e)
}

func (a *App) RecordMetric(name string, value float64) {
	a.dna.Record(name, value)
}

func (a *App) IsAnomaly(metric string, value float64) bool {
	return a.dna.IsAnomaly(metric, value)
}

func (a *App) HealthScore(current map[string]float64) float64 {
	return a.dna.HealthScore(current)
}

func (a *App) Start() {
	a.running = true
	a.bus.Subscribe("*", func(e *event.Event) {
		a.healer.Handle(e)
	})
}

func (a *App) Stop()           { a.running = false }
func (a *App) IsRunning() bool { return a.running }
func (a *App) Config() Config  { return a.config }

func MatchSeverity(min event.Severity) healing.MatchFunc { return healing.MatchSeverity(min) }
func MatchSource(source string) healing.MatchFunc        { return healing.MatchSource(source) }
func MatchContains(substr string) healing.MatchFunc      { return healing.MatchContains(substr) }
func ActionExec(cmd string) healing.ActionFunc           { return healing.ActionExec(cmd) }

File: sdk/typescript/src/immortal.ts
import { ImmortalEvent, Severity } from "./event";
import * as crypto from "crypto";

export type HealingMode = "ghost" | "reactive" | "predictive" | "autonomous";

export interface HealingRule {
  name: string;
  match: (event: ImmortalEvent) => boolean;
  action: (event: ImmortalEvent) => Promise<void>;
}

export interface ImmortalConfig {
  name?: string;
  mode?: HealingMode;
  endpoint?: string;
  ghostMode?: boolean;
}

export class Immortal {
  private config: Required<ImmortalConfig>;
  private rules: HealingRule[] = [];
  private running = false;
  private events: ImmortalEvent[] = [];

  constructor(config: ImmortalConfig = {}) {
    this.config = {
      name: config.name ?? "immortal-app",
      mode: config.mode ?? "reactive",
      endpoint: config.endpoint ?? "http://localhost:7777",
      ghostMode: config.ghostMode ?? false,
    };
  }

  addRule(rule: HealingRule): void {
    this.rules.push(rule);
  }

  heal(rule: {
    name: string;
    when: (event: ImmortalEvent) => boolean;
    do: (event: ImmortalEvent) => Promise<void>;
  }): void {
    this.addRule({
      name: rule.name,
      match: rule.when,
      action: rule.do,
    });
  }

  async handleEvent(event: ImmortalEvent): Promise<string[]> {
    this.events.push(event);
    const matched: string[] = [];

    for (const rule of this.rules) {
      if (rule.match(event)) {
        matched.push(rule.name);
        if (!this.config.ghostMode) {
          await rule.action(event);
        }
      }
    }

    return matched;
  }

  start(): void {
    this.running = true;
  }

  stop(): void {
    this.running = false;
  }

  isRunning(): boolean {
    return this.running;
  }

  getConfig(): Required<ImmortalConfig> {
    return { ...this.config };
  }

  getEvents(): ImmortalEvent[] {
    return [...this.events];
  }
}
